[Bug]: Panic: runtime error: invalid memory address or nil pointer dereference during UpgradeResourceState

### Github handle

VerbekeJarne

### Please describe the issue you are facing?

I'm testing the upgrade from version `~> 1.2` to version `~> 2.2`. While doing that and running a `terraform plan` I get:

```
Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).UpgradeResourceState call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-terracurl_v2.2.0 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x105142edc]

goroutine 52 [running]:
github.com/devops-rob/terraform-provider-terracurl/internal/provider.(*CurlResource).UpgradeState.func1({0x1057b7a08, 0x14000326750}, {0x140000abdc0?, 0x0?}, 0x140001380f0)
        /Users/rbarnes/Developer/devops-rob/terraform-provider-terracurl/internal/provider/curl_resource.go:959 +0x6c
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).UpgradeResourceState(0x14000226308, {0x1057b7a08, 0x14000326750}, 0x1400054e600, 0x140003e9648)
        /Users/rbarnes/go/pkg/mod/github.com/hashicorp/terraform-plugin-framework@v1.17.0/internal/fwserver/server_upgraderesourcestate.go:205 +0x734
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).UpgradeResourceState(0x14000226308, {0x1057b7a08?, 0x14000326660?}, 0x140000abde0)
        /Users/rbarnes/go/pkg/mod/github.com/hashicorp/terraform-plugin-framework@v1.17.0/internal/proto6server/server_upgraderesourcestate.go:51 +0x23c
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).UpgradeResourceState(0x140003021e0, {0x1057b7a08?, 0x14000301710?}, 0x14000189360)
        /Users/rbarnes/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.29.0/tfprotov6/tf6server/server.go:800 +0x1f0
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_UpgradeResourceState_Handler({0x105791700, 0x140003021e0}, {0x1057b7a08, 0x14000301710}, 0x140001bda00, 0x0)
        /Users/rbarnes/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.29.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:699 +0x1b8
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000304248, {0x1057b7a08, 0x1400054e0f0}, 0x1400013e680, 0x140003015f0, 0x105db5750, 0x0)
        /Users/rbarnes/go/pkg/mod/google.golang.org/grpc@v1.78.0/server.go:1428 +0xc94
google.golang.org/grpc.(*Server).handleStream(0x14000304248, {0x1057b8478, 0x1400048a000}, 0x1400013e680)
        /Users/rbarnes/go/pkg/mod/google.golang.org/grpc@v1.78.0/server.go:1832 +0x9e0
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        /Users/rbarnes/go/pkg/mod/google.golang.org/grpc@v1.78.0/server.go:1063 +0x74
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 23
        /Users/rbarnes/go/pkg/mod/google.golang.org/grpc@v1.78.0/server.go:1074 +0x120

Error: The terraform-provider-terracurl_v2.2.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

2026-02-13T15:14:06.018+0100 [DEBUG] provider: plugin exited
```

### Please detail the steps to reproduce the issue?

1. Test the upgrade with below code (running Terraform version 1.13.2) :

```
terraform {
  required_version = "~> 1.13.2"
  required_providers {
    terracurl = {
      source  = "devops-rob/terracurl"
      version = "~> 1.2"
    }
  }
}
resource "terracurl_request" "get_example" {
  name   = "example-get-call"
  url    = "https://jsonplaceholder.typicode.com/posts/1"
  method = "GET"
  headers = {
    "Content-Type" = "application/json"
  }
  response_codes     = [200, 201]
}
```

All works fine:
```
Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

2. Now change the version from `~> 1.2` to `~> 2.2` and run `terraform init -upgrade && terraform plan` which will show:

```
Planning failed. Terraform encountered an error while generating this plan.
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).UpgradeResourceState call. The plugin logs may contain more details.
╵
Stack trace from the terraform-provider-terracurl_v2.2.0 plugin:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x105142edc]
goroutine 52 [running]:
github.com/devops-rob/terraform-provider-terracurl/internal/provider.(*CurlResource).UpgradeState.func1({0x1057b7a08, 0x14000326750}, {0x140000abdc0?, 0x0?}, 0x140001380f0)
        /Users/rbarnes/Developer/devops-rob/terraform-provider-terracurl/internal/provider/curl_resource.go:959 +0x6c
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).UpgradeResourceState(0x14000226308, {0x1057b7a08, 0x14000326750}, 0x1400054e600, 0x140003e9648)
        /Users/rbarnes/go/pkg/mod/github.com/hashicorp/terraform-plugin-framework@v1.17.0/internal/fwserver/server_upgraderesourcestate.go:205 +0x734
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).UpgradeResourceState(0x14000226308, {0x1057b7a08?, 0x14000326660?}, 0x140000abde0)
        /Users/rbarnes/go/pkg/mod/github.com/hashicorp/terraform-plugin-framework@v1.17.0/internal/proto6server/server_upgraderesourcestate.go:51 +0x23c
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).UpgradeResourceState(0x140003021e0, {0x1057b7a08?, 0x14000301710?}, 0x14000189360)
        /Users/rbarnes/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.29.0/tfprotov6/tf6server/server.go:800 +0x1f0
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_UpgradeResourceState_Handler({0x105791700, 0x140003021e0}, {0x1057b7a08, 0x14000301710}, 0x140001bda00, 0x0)
        /Users/rbarnes/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.29.0/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:699 +0x1b8
google.golang.org/grpc.(*Server).processUnaryRPC(0x14000304248, {0x1057b7a08, 0x1400054e0f0}, 0x1400013e680, 0x140003015f0, 0x105db5750, 0x0)
        /Users/rbarnes/go/pkg/mod/google.golang.org/grpc@v1.78.0/server.go:1428 +0xc94
google.golang.org/grpc.(*Server).handleStream(0x14000304248, {0x1057b8478, 0x1400048a000}, 0x1400013e680)
        /Users/rbarnes/go/pkg/mod/google.golang.org/grpc@v1.78.0/server.go:1832 +0x9e0
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        /Users/rbarnes/go/pkg/mod/google.golang.org/grpc@v1.78.0/server.go:1063 +0x74
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 23
        /Users/rbarnes/go/pkg/mod/google.golang.org/grpc@v1.78.0/server.go:1074 +0x120
Error: The terraform-provider-terracurl_v2.2.0 plugin crashed!
This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
2026-02-13T15:14:06.018+0100 [DEBUG] provider: plugin exited
```

### TerraCurl provider version

2.2.0

### What version of Terraform are you using?

1.13.2

### What Operating system architecture are you running?

darwin_arm64

### Relevant log output

```shell

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,9 @@
+## 2.5.3
+
+BUG FIXES:
+
+- Fix panic upgrading `terracurl_request` state from provider v1.x when `UpgradeResourceState` receives nil prior state. Closes #134.
+
 ## 2.5.2
 
 BUG FIXES:

--- a/internal/provider/curl_resource.go
+++ b/internal/provider/curl_resource.go
@@ -1020,13 +1020,24 @@ func (r *CurlResource) UpgradeState(ctx context.Context) map[int64]resource.Stat
 
 				var oldState CurlResourceModel
 
-				// First, try the normal state extraction
-				diags := req.State.Get(ctx, &oldState)
+				var diags diag.Diagnostics
+				useRawState := req.State == nil
 
-				if diags.HasError() {
-					tflog.Info(ctx, "Direct state extraction failed, attempting to extract values from raw state", map[string]interface{}{
-						"errors": diags.Errors(),
-					})
+				if !useRawState {
+					diags = req.State.Get(ctx, &oldState)
+					if diags.HasError() {
+						useRawState = true
+					}
+				}
+
+				if useRawState {
+					if req.State == nil {
+						tflog.Info(ctx, "Prior state is nil, attempting to extract values from raw state")
+					} else {
+						tflog.Info(ctx, "Direct state extraction failed, attempting to extract values from raw state", map[string]interface{}{
+							"errors": diags.Errors(),
+						})
+					}
 
 					// Try to extract what we can from the raw state
 					if req.RawState != nil && len(req.RawState.JSON) > 0 {

--- a/internal/provider/curl_resource_test.go
+++ b/internal/provider/curl_resource_test.go
@@ -1762,3 +1762,86 @@ func TestCurlResource_StateUpgrade_EmptyDestroyParameters(t *testing.T) {
 		t.Errorf("Expected id to be preserved, got '%s'", upgradedState.Id.ValueString())
 	}
 }
+
+func TestCurlResource_StateUpgrade_NilRequestState(t *testing.T) {
+	ctx := context.Background()
+	r := &CurlResource{}
+
+	upgraders := r.UpgradeState(ctx)
+	upgrader, ok := upgraders[0]
+	if !ok {
+		t.Fatal("No upgrader found for version 0")
+	}
+
+	// v1.2-style state JSON from issue #134 (provider SDK -> framework upgrade).
+	rawStateJSON := `{
+		"id": "example-get-call",
+		"name": "example-get-call",
+		"url": "https://jsonplaceholder.typicode.com/posts/1",
+		"method": "GET",
+		"headers": {"Content-Type": "application/json"},
+		"response_codes": ["200", "201"],
+		"response": "{\"userId\":1,\"id\":1,\"title\":\"sunt aut facere\",\"body\":\"quia et suscipit\"}",
+		"status_code": "200"
+	}`
+
+	rUpgrade := &CurlResource{}
+	schemaResp := &resource2.SchemaResponse{}
+	rUpgrade.Schema(ctx, resource2.SchemaRequest{}, schemaResp)
+	schemaVar := schemaResp.Schema
+
+	req := resource2.UpgradeStateRequest{
+		State: nil,
+		RawState: &tfprotov6.RawState{
+			JSON: []byte(rawStateJSON),
+		},
+	}
+
+	resp := &resource2.UpgradeStateResponse{
+		State: tfsdk.State{
+			Schema: schemaVar,
+		},
+	}
+
+	upgrader.StateUpgrader(ctx, req, resp)
+	if resp.Diagnostics.HasError() {
+		t.Fatalf("upgrade failed: %v", resp.Diagnostics)
+	}
+
+	var upgradedState CurlResourceModel
+	diags := resp.State.Get(ctx, &upgradedState)
+	if diags.HasError() {
+		t.Fatalf("error getting upgraded state: %v", diags)
+	}
+
+	if upgradedState.Id.ValueString() != "example-get-call" {
+		t.Errorf("Expected id 'example-get-call', got '%s'", upgradedState.Id.ValueString())
+	}
+	if upgradedState.Name.ValueString() != "example-get-call" {
+		t.Errorf("Expected name 'example-get-call', got '%s'", upgradedState.Name.ValueString())
+	}
+	if upgradedState.Url.ValueString() != "https://jsonplaceholder.typicode.com/posts/1" {
+		t.Errorf("Expected url preserved, got '%s'", upgradedState.Url.ValueString())
+	}
+	if upgradedState.Method.ValueString() != "GET" {
+		t.Errorf("Expected method 'GET', got '%s'", upgradedState.Method.ValueString())
+	}
+	if upgradedState.Headers.IsNull() {
+		t.Error("Expected headers to be preserved")
+	}
+	if upgradedState.ResponseCodes.IsNull() {
+		t.Error("Expected response_codes to be preserved")
+	}
+	if !upgradedState.SkipRead.ValueBool() {
+		t.Error("Expected skip_read to be set to true in v1 upgrade")
+	}
+	if !upgradedState.ReadUrl.IsNull() {
+		t.Error("Expected read_url to be null in v1 upgrade")
+	}
+	if !upgradedState.ReadResponseCodes.IsNull() {
+		t.Error("Expected read_response_codes to be null in v1 upgrade")
+	}
+	if upgradedState.ResponseSensitive.IsNull() || upgradedState.ResponseSensitive.ValueBool() {
+		t.Error("Expected response_sensitive to default to false")
+	}
+}

